WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift

```diff
@@ -0,0 +1,172 @@
+import Foundation
+import StoreKit
+
+protocol InAppFeedbackPromptPresenting {
+    func presentIfNeeded(in controller: UIViewController) -> Bool
+}
+
+struct InAppFeedbackPromptCoordinator: InAppFeedbackPromptPresenting {
+
+    // MARK: - Properties
+
+    private let appRatingUtility: AppRatingUtility
+    private let userDefaults: UserDefaults
+
+    // MARK: - Init
+
+    init(utility: AppRatingUtility = .shared, userDefaults: UserDefaults = .standard) {
+        self.userDefaults = userDefaults
+        self.appRatingUtility = utility
+    }
+
+    // MARK: - Showing the Prompt
+
+    func shouldShowPromptForAppReview() -> Bool {
+        return appRatingUtility.shouldPromptForAppReview()
+    }
+
+    func presentIfNeeded(in controller: UIViewController) -> Bool {
+        guard shouldShowPromptForAppReview() else {
+            return false
+        }
+        let alert = self.feedbackAlert(in: controller)
+        controller.present(alert, animated: true)
+        appRatingUtility.userWasPromptedToReview()
+        WPAnalytics.track(.appReviewsSawPrompt)
+        return true
+    }
+
+    private func feedbackAlert(in controller: UIViewController) -> UIAlertController {
+        let alert = UIAlertController(
+            title: Strings.FeedbackAlert.title,
+            message: Strings.FeedbackAlert.message,
+            preferredStyle: .alert
+        )
+        let yes = UIAlertAction(title: Strings.FeedbackAlert.yes, style: .default) { _ in
+            self.handlePositiveFeedback(in: controller)
+        }
```

Comment:
I suppose there's no retain cycle as `UIAlertController` will be deallocated once the function ends. Though I thought it could make it less prone for a cycle in the future. Up to you.

Reply:
Good catch, I'll add `[weak self]` to this closure and all other UIAlertAction closures.

Reply:
I got this error when adding `[weak self]`:
```
'weak' may only be applied to class and class-bound protocol types, not 'InAppFeedbackPromptCoordinator'
```
That makes sense since `InAppFeedbackPromptCoordinator` is a `struct`.

Reply:
Oh, that actually may not be such a great idea as it contains other reference types. So when you pass it around, its value type properties will be copied but the reference types will be modified throughout every copy.